### PR TITLE
mapgen: simplify and fix remove_vehicles

### PR DESCRIPTION
**Pull request by andrei8l**
#### Summary
None

#### Purpose of change
The function for `"remove_vehicles"` is overly complicated
Both `"remove_vehicles"` and `"remove_all"` clear the main vehicle cache unconditionally
Both also crash the game if the nested/update mapgen runs in bounds of the main map
`map::load()` changes the (relative) position of the vehicles which makes it harder to fix the above issues
update mapgen heavy-handedly emulates a long-range teleport to reset caches and trigger traps

Fixes: #58285
Fixes: #57581
Fixes: #57717 (note: this one cannot be reproduced directly since #58112)
(Probably) Fixes: #55176
(Probably) Fixes: #58267 these two are likely the same as the previous two but don't have full reproduction steps. The bug descriptions are nearly the same though
Fixes: #55799 you get some message spam now but at least vehicles work through mapgen updates

#### Describe the solution
Simplify `jmapgen_remove_vehicles`
Reset vehicle caches on the main map if a map was loaded in bounds

TODO:
- [x] unit test

#### Describe alternatives you've considered
N/A

#### Testing
Unit tests

<details>
<summary>Additional manual testing</summary>

Spawn two cars on one submap then spawn a `pond_6x6` nested map that overlaps only one of them. There should be no crashes and only the overlapping car should disappear.

Spawn a `fire_engine` and a `wheelchair` then spawn `mapgen_test_nested_remove_veh` on the same submap. only the `fire_engine` should disappear.

Load the save from https://github.com/CleverRaven/Cataclysm-DDA/issues/58285#issuecomment-1159188755 and wait one turn. The military cargo truck should still be there. Walk down/teleport to the nearby `roadstop_a` and verify that the caravaner truck was removed.

Drive vehicles inside the color-coded zones of the TCL to trigger map updates.
</details>

#### Additional context
@jbytheway please take a look and let me know if there's a better solution off the top of your head.

### CODE REVIEW

File: src/mapgen.cpp [hunk 1]
```diff
@@ -3006,38 +3006,29 @@ class jmapgen_remove_vehicles : public jmapgen_piece
...
-                        }
-                        if( remove ) {
-                            dat.m.destroy_vehicle( &vp->vehicle() );
-                            get_map().clear_vehicle_level_caches();
-                        }
+                        removed = true;
                     }
                 }
             }
+            if( removed and main_inbounds ) {
```
Comment (jbytheway) [inline src/mapgen.cpp:3027]:
```suggestion
            if( removed && main_inbounds ) {
```
We prefer the non-word operators.

File: src/mapgen.cpp [hunk 2]
```diff
@@ -3080,17 +3071,27 @@ class jmapgen_remove_all : public jmapgen_piece
...
                 if( optional_vpart_position vp = dat.m.veh_at( p ) ) {
+                    if( main_inbounds ) {
+                        get_map().remove_vehicle_from_cache( &vp->vehicle(), start.z, end.z );
+                    }
                     dat.m.destroy_vehicle( &vp->vehicle() );
-                    get_map().clear_vehicle_level_caches();
+                    removed = true;
                 }
             }
+            if( removed and main_inbounds ) {
```
Comment (jbytheway) [inline src/mapgen.cpp:3090]:
```suggestion
            if( removed && main_inbounds ) {
```

**Review by jbytheway — COMMENTED** [review]
Thanks for finding a way to resolve this.  It's been bugging me for a while but I never had the willpower to actually look for a solution.  This approach looks reasonable.  Certainly a unit test would be welcome, since this is a tricky issue to understand and one that might regress.

**Comment by andrei8l** [comment]
Thanks for taking a look. I've moved the vehicle-pos-fixing bit to `~map()` because otherwise vehicles are still broken after mapgen is done. The proper solution is to get rid of `vehicle::sm_pos` but I imagine there's a kraken of legacy code behind it.

I'll get the test units sorted soon.

**Comment by andrei8l** [comment]
I believe this is ready, but I'm not entirely happy with how the test unit works.

As a bonus though, performance inside the TCL should be a bit better. Probably f*xes (or alleviates) #57521, #55177, and #55019 but I haven't playtested enough to be sure.

**Comment by andrei8l** [comment]
Thanks for merging!

**Pull request merged**